messagecatalog/spec.md

```diff
@@ -0,0 +1,1475 @@
+posted with some or all of its versions, the versions are held in an array named
+`versions` and all common attributes are held in the containing object.
+
+The example below, illustrating a CloudEvent without payload for brevity, shows an
+example of a _message definition version_ document:
+
+```JSON
+{
+    "id" : "io.cloudevents.messagecatalog.deleted",
+    "authority" : "cloudevents.io",
+    "description" : "Raised when an object has been deleted from the catalog",
+    "format" : "CloudEvents",
+    "versions" : [
+      {
+        "version" : 1,
+        "definition" : {
+          "attributes" : {
+            "specversion" : {
+              "value" :"1.0",
+              "required" : true,
+              "description" : "CloudEvents specification version"
+            },
+            "type" : {
+              "value" : "io.cloudevents.messagecatalog.deleted",
+              "required" : true,
+              "description" : "type of this event"
+            }
+          }
+        }
+    },
+    {
+        "version" : 2,
+        "definition" : {
+          "attributes" : {
+            "specversion" : {
+              "value" :"1.0",
+              "required" : true,
+              "description" : "CloudEvents specification version"
+            },
+            "type" : {
+              "value" : "io.cloudevents.messagecatalog.deleted",
+              "required" : true,
+              "description" : "type of this event"
+            },
+            "subject" : {
+              "required" : true,
+              "description" : "id of the deleted event definition"
+            }
+          }
+        }
+    }
+  ]
+}
+```
+
+For version "2" of the above definition, this is the standalone _message
+definition version_ document:
+
+``` JSON
+{
+  "id" : "io.cloudevents.messagecatalog.deleted",
+  "authority" : "cloudevents.io",
+  "description" : "Raised when an object has been deleted from the catalog",
+  "format" : "CloudEvents",
+  "version" : 2,
+  "definition" : {
+    "attributes" : {
+      "specversion" : {
+        "value" :"1.0",
+        "required" : true,
+        "description" : "CloudEvents specification version"
+      },
+      "type" : {
+        "value" : "io.cloudevents.messagecatalog.deleted",
+        "required" : true,
+        "description" : "type of this event"
+      },
+      "subject" : {
+        "required" : true,
+        "description" : "id of the deleted event definition"
+      }
+    }
+  }
+}
+```
+
+The API is created such that a message definition MUST be created by posting an
+initial version and that the message definition (including any common parts like
+the documentation) MUST only be changed by posting a new version. Existing
+versions can be individually retrieved and deleted, but not modified.
+
+A newer message definition version might introduce breaking changes or it might
+only introduce careful changes that preserve compatibility. These strategies are
```

Comment:
semver passes a tool to the implementors to make smarter choices on how to upgrade and represent versions.

Reply:
or you can do what proto does and deprecate  fields and never delete them